### DISCUSSION TITLE
Export Map server side to PNG

### DISCUSSION THREAD

**Discussion by Cayllen** [openlayers/openlayers#15886] [Q&A]
Hey all,
I'm hitting a bit of a wall here, but how would I go about exporting the map server side and getting the image data? 
Have not found any useful hints yet how I could go about this

**Comment by ahocevar** [comment]
You may want to take a look at https://github.com/ahocevar/ol-tileserver for inspiration.

**Reply by Cayllen** [reply]
Thanks! Will check it out. I was looking for an easy solution, but it seems there is not really anything available. I went now for a **very** dirty puppeteer solution but seems to work actually quite good. The code in case anyone also needs it. Param1+Param2 is used to bring some objects from the outside into puppeteer. Those have to be in JSON/GeoJSON and should not be an actual OL object (then it didn't work for me at least)

```js
(async () => {
		try {
			const browser = await puppeteer.launch({ args: ['--no-sandbox', '--disable-setuid-sandbox'] });
			const page = await browser.newPage();
			await page.setViewport({ width: 900, height: 700 });
			await page.setContent(
				`<html><body style="margin:0;padding:0">
                 <div id="map" style="height:100%; width:100%; position:absolute"></div></body></html>`
			);
			await page.addScriptTag({ url: 'https://cdn.jsdelivr.net/npm/ol@v9.2.4/dist/ol.js' });
			await page.evaluate(
				async (param1, param2) => {
					const map = new ol.Map({
						target: 'map',
						layers: [
							new ol.layer.Tile({
								source: new ol.source.OSM()
							}),
						],
						view: new ol.View({
							center: ol.proj.fromLonLat([0,0]),
							zoom: 3
						}),
						controls: []
					});
				},
				param1,
				param2
			);

			await new Promise((r) => setTimeout(r, 1200));

			await page.screenshot({
				path: 'packages/core/src/screenshot.jpg'
			});
			await browser.close();
		} catch (error) {
			console.error(error);
		}
	})();
```
